lib/cn/cn_tree.c

```diff
@@ -2038,21 +2070,31 @@ cn_comp_spill(struct cn_compaction_work *w)
         while ((ss = cn_subspill_pop(node))) {
             err = cn_subspill_apply(ss);
             if (err)
-                break;
+                goto errout;
 
             atomic_inc(&node->tn_sgen);
+            node->tn_ss_spilling--;
         }
+
+        spillingp = NULL;
     }
 
     w->cw_t3_build = get_time_ns();
 
-    if (err && merr_errno(err) != ESHUTDOWN)
-        kvdb_health_error(tree->ct_kvdb_health, err);
+  errout:
+    if (err) {
+        if (merr_errno(err) != ESHUTDOWN)
+            kvdb_health_error(tree->ct_kvdb_health, err);
 
-    /* Serialize the deletion of input kvsets.
-     */
-    if (!err && cn_node_spill_wait(w, w->cw_tree->ct_root))
-        cn_spill_delete_kvsets(w);
+        if (spillingp)
+            (*spillingp) -= 1;
+    } else {
+
+        /* Serialize the deletion of input kvsets.
+         */
+        if (cn_node_spill_wait(w, w->cw_tree->ct_root))
+            cn_spill_delete_kvsets(w);
```

Comment:
This approach to cleaning up kvsets from root node is causing me grief.  Can we not do it when the tree is updated after the spilling to the last leaf?   We already let compaction jobs commit the leaf work of other jobs (see `cn_subspill_pop`), why not let the last job do the final cleanup?  Then we can get rid of:
- tying up threads waiting here
- functions `cn_node_spill_wait()` and `cn_node_spill_broadcast()`
- the spill wait mutex and condition var

@gsramdasi thoughts?

Reply:
It would be good if rspill threads that have to wait here could instead dump that work off on the thread they're waiting for, then move on to pick up new work...

Reply:
@alexttx  Yes, that would be cleaner. But since this was well tested, I did not change that code. I can look at it later.